#the following R script has been designed to carry out deseq2 on count files
#generated by HTSeq-count. The script was designed to work on files in the
#format conditionname_sampleID_countsy.txt where y was a number


#Install required packages if needed 
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("DESeq2", force = TRUE)
install.packages("dplyr", dependencies = TRUE)
install.packages("plyr", dependencies = TRUE)


library("DESeq2")

#set working directory
setwd("/Users/dasithperera/rnaseq_analysis/083022/cat/htseq/")

#sanity check of wd
directory <- getwd()
directory

#obtain sample filesnames by searching for count files in wd  
sampleFiles <- grep ("count" ,list.files(directory),value=TRUE)

#extract sample condition information by removing everything after the first underscore. eg. healthy_xx1_counts1.txt will extract healthy
sampleCondition <- sub("_.*", "", sampleFiles)

#extract sample name by capturing information between underscores
sampleID <- sub("^[^_]+_([^_]+)_.*", "\\1", sampleFiles)

#generate table for use in deseq from HTSeqcount files
sampleTable <- data.frame(sampleName = sampleID,
                          fileName = sampleFiles,
                          condition = sampleCondition)

#create deseq data object from above table and HTSeqcount files
ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = directory,
                                       design= ~condition)

#run deseq2 analysis
dds <- DESeq(ddsHTSeq)

#get results from deseq2
res <- results(dds)
res

library("ggplot2")
library("EnhancedVolcano")

#using rlog transformed data generate and save pca_plot in svg format
rld = rlogTransformation(dds)
svg("./pcaplot.svg")
plotPCA(rld, intgroup=c("condition"))
dev.off()

#generate a volcano plot and export in svg format
svg("./volcanoplot.svg")
EnhancedVolcano(res,
                lab = NA,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-5, 8))
dev.off()


#creating a column called id and moving the rownames i.e. the gene ids to that column for data wrangling 
res$id <- rownames(res)
rownames(res) <- NULL

#exporting raw csv file of deseq results
write.csv(res, file = "./raw_deseqtable.csv")

#inputing raw csv file 
deseqtable <- read.csv("./raw_deseqtable.csv")

#converting log2foldchange to real fold change and adding the relevant column 
deseqtable$fold_change <- 2^deseqtable[,3]


#find gff file
gff <- dir(pattern="\\.gff$")

#reading gff 
tab <- read.table(gff, sep="\t", header=FALSE, comment.char="#",
                  na.strings=".", stringsAsFactors=FALSE,
                  quote="", fill=FALSE)

#importing splitstackshape as gff file requires parsing 
library(splitstackshape)


#creating filtered gff file to contain only CDS entries as htseq only mapped to CDS
filt_tab <- tab[tab$V3 == "CDS", ]


#separating column V9 by = and then ; to extract gene ID that matches format from deseq output
results2 <- concat.split(data = filt_tab, split.col = "V9", sep = "=", drop = TRUE)
gff_file <- concat.split(data = results2, split.col = "V9_02", sep = ";", drop = TRUE)

#adding headers to the gene id column that should match up with ids from deseq results
names(gff_file)[26] <- "id"
#adding header to the column containing gene information 
names(gff_file)[18] <- "gff_annotation"

#filtering to create a table containing only the above columns
filt_gff <- gff_file[,c(26, 18)]

#importing dplyr
library("dplyr")

#merging deseq results with the above filtered gff file to provide gene annotations of DEG  
merged_gff_deseq<- inner_join(deseqtable, filt_gff, by="id")

#creating filtered table to contain fold change, gene id, pvalues, padjusted, gff annotation 
filtered_merged_gff_deseq <- merged_gff_deseq[,c(9, 8, 6, 7, 3, 10)]

#exporting filtered file as a csv
write.csv(filtered_merged_gff_deseq, file = "./rnaseq_analysis/raw_condition2_vs_condition1.csv")

#filtering above based on padjusted value <0.05
sig <- filter(filtered_merged_gff_deseq, padj<0.05)

#taking padjusted rows and extracting only the significantly upregulated genes i.e. fold change >2
sigup <- filter(sig, fold_change>2)

#taking padjusted rows and extracting only the significantly downregulated genes i.e. fold change <2
sigdown <- filter(sig, fold_change<0.5)

#converting fold change of downregulated genes from fractions of expression to negative fold change e.g 0.5 to -2 
colnames(sigdown)
sigdown$fold_change <- -1/sigdown[,1]

#sorting by greatest fold change
sorted_sigup<- arrange(sigup, desc(fold_change))
sorted_sigdown<- arrange(sigdown, fold_change)

#exporting all significantly expressed genes i.e. padj<0.05
write.csv(sig, file = "./rnaseq_analysis/allsig_conditon2_vs_condition1.csv")

#exporting significantly upregulated genes i.e. padj<0.05 and fold change >2
write.csv(sorted_sigup, file = "./rnaseq_analysis/upsig_conditon2_vs_condition1.csv")

#exporting significantly upregulated genes i.e. padj<0.05 and fold change <-2
write.csv(sorted_sigdown, file = "./rnaseq_analysis/downsig_conditon2_vs_condition1.csv")
